README.rst
.. image:: https://raw.githubusercontent.com/neo4j-contrib/neomodel/master/doc/source/_static/neomodel-300.png
   :alt: neomodel

An Object Graph Mapper (OGM) for the neo4j_ graph database, built on the awesome neo4j_driver_

- Familiar Django model style definitions.
- Powerful query API.
- Enforce your schema through cardinality restrictions.
- Full transaction support.
- Thread safe.
- pre/post save/delete hooks.
- Django integration via django_neomodel_

.. _django_neomodel: https://github.com/neo4j-contrib/django-neomodel
.. _neo4j: https://neo4j.com/
.. _neo4j_driver: https://github.com/neo4j/neo4j-python-driver

.. image:: https://secure.travis-ci.org/neo4j-contrib/neomodel.png
    :target: https://secure.travis-ci.org/neo4j-contrib/neomodel/

.. image:: https://readthedocs.org/projects/neomodel/badge/?version=latest
    :alt: Documentation Status
    :scale: 100%
    :target: https://neomodel.readthedocs.io/en/latest/?badge=latest


Documentation
=============

Available on readthedocs_.

.. _readthedocs: http://neomodel.readthedocs.org

Requirements
============

- Python 3.5+
- `Supported version <https://neo4j.com/developer/kb/neo4j-supported-versions/>`_ of neo4j


Installation
============

Install from pypi (recommended)::

    $ pip install neomodel

To install from github::

    $ pip install git+git://github.com/neo4j-contrib/neomodel.git@HEAD#egg=neomodel-dev

Upgrading 3.x to 4.x
====================

 * The deprecated search() method on RelationshipManager has been removed (use filter() instead)
 * The deprecated NormalProperty has been removed (use NormalizedProperty instead)
 * The "bolt+routing" scheme has been replaced with "neo4j". Using "bolt+routing" produces a deprecation warning, and replaces the scheme with "neo4j"
 * Neo4j no longer supports Python 2.7, or versions below 3.5
 * neobolt and neotime are no longer required
 * The ConnectionExpired exception has been removed
 * Be aware of changes in https://github.com/neo4j/neo4j-python-driver/wiki/4.0-changelog
    * Connections are now unencrypted by default; to reproduce former behaviour, set ENCRYPTED_CONNECTION to True and scheme to "neo4j+s", "neo4j+ssc", "bolt+s" or "bolt+ssc"
    * The neo4j.v1 subpackage is now no longer available; all imports should be taken from the neo4j package instead.
    * CypherError is now Neo4jError
    * max_pool_size is now max_connection_pool_size
    * access_mode is now default_access_mode

Using an older version
======================

To use an older version, for instances where you have to use Python 2.7, use neo4j-driver 1.7 and neomodel 3.3.2.


Contributing
============

Ideas, bugs, tests and pull requests always welcome. 

As of release `4.0.0` we now have a curated list of issues / development targets for
`neomodel` available on `the Wiki <https://github.com/neo4j-contrib/neomodel/wiki/TODOs-&-Enhancements>`_.

If you are interested in developing `neomodel` further, pick a subject from the list and open a Pull Request (PR) for 
it. If you are adding a feature that is not captured in that list yet, consider if the work for it could also 
contribute towards delivering any of the existing todo items too.

Running the test suite
----------------------

Make sure you have a Neo4j database version 3 or higher to run the tests on.::

    $ export NEO4J_BOLT_URL=bolt://neo4j:neo4j@localhost:7687 # (the default)

Setup a virtual environment, install neomodel for development and run the test suite::

    $ virtualenv venv
    $ source venv/bin/activate
    $ python setup.py develop
    $ pytest

If you are running a neo4j database for the first time the test suite will set the password to 'test'.
If the database is already populated, the test suite will abort with an error message and ask you to re-run it with the
`--resetdb` switch. This is a safeguard to ensure that the test suite does not accidentally wipe out a database if
you happen to not have restarted your Neo4j server to point to a (usually named) `debug.db` database.

If you have ``docker-compose`` installed, you can run the test suite against all supported Python
interpreters and neo4j versions::

    # in the project's root folder:
    $ ./tests-with-docker-compose.sh


.. image:: https://badges.gitter.im/Join%20Chat.svg
   :alt: Join the chat at https://gitter.im/neo4j-contrib/neomodel
   :target: https://gitter.im/neo4j-contrib/neomodel?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge


Changelog
Version 4.0.0 2020-03-07
* Updated to work with Neo4j-driver 4.0 (#489) - Jais Vemmer
* Fixed Create Unique no longer being supported (#488) - Jais Vemmer
* Fixed {var} style variables no longer being supported (#487) - Jais Vemmer
* Fixed neo4j:// URI validity (#486) - Jais Vemmer
* Added support for Neo4j 4.0 (#485) - Jais Vemmer

Version 3.3.2 2019-09-29
* Fixed validation for unique properties to also be optional (#470) - Jon Daly
* Fixed tests-with-docker-compose.sh (#452) - Jorge Valhondo Rama
* Added exclusion example to documentation and improved naming consistency (#456, #466) - Elena Williams 
* Fixed the Travis-CI build problems by switching to openjdk (#471) - Duncan Booth
* Minor revisions in `test_set_connection`, `spatial datatypes` and documentation (#442, #446, #447) - Athanasios Anastasiou
* Added `max_length` constraint on `StringProperty` (#445) - Lazy-Y
* Upgraded neo4j python driver requirement to 1.7.2 (#432) - Robert Grant
* Added a `DateTimeFormatProperty` (#428) - Yu Shengnan
* Updated `setup.py` so that it excludes the `test/` and `test/test_contrib/` sub-packages (#426) - Jorge Valhondo Rama
* Updated getting_started.rst typo (#419) - fredthehead
* Fixed NeomodelPoint instantiation bug (#418) - Athanasios Anastasiou

Version 3.3.1 2019-02-08
 * Fixed a number of warnings due to deprecations both within neomodel and pytest and overall improvements in
   code style (#381) - Abhishek Modi
 * Added support for spatial data through neomodel.contrib.spatial_datatypes (#384) - Athanasios Anastasiou
 * Added the ability to filter "left-hand statements" (in NodeSet expressions) too (#395) - Grzegorz Grzywacz
 * Refactor the Node Class Registry to make util.Database a true Singleton (#403) - Giorgos Oikonomou
     * Many thanks to Giorgos Oikonomou, Jon Daly, Adam Romano, Andrew Tergis, Mato Žgajner, Mostafa Moradian, mjmare,
       Phoebe Bright, Robert Grant, jberends and anyone else who helped flag, track and correct this bug. For more
       information, please see: https://github.com/neo4j-contrib/neomodel/issues/378
     * Added the ClassAlreadyDefined exception to prevent against accidental redefinitions of data model classes (#409)
       - Athanasios Anastasiou
 * Added the ability to lazily `.nodes.get()` and `.nodes.all()` (#406) - Matan Hart
 * Fixed a bug in the assumed direction of relationships in _build_merge_query (#408) - MrAnde7son

Version 3.3.0 2018-10-04
 * Added support for Q() in filter and exclude (#360) - Juan H. Hidalgo
 * Added  object docs and examples - Juan H. Hidalgo
 * Raise DoesNotExist when refreshing a non existent inflated node (#355) - lerela
 * Empty filter bug - Mardanov Timur Rustemovich
 * Closed #361 Nodes connected with two or more relationships - Mardanov Timur Rustemovich
 * Exclude method fixed - Mardanov Timur Rustemovich
 * Filter with OR fixed - Mardanov Timur Rustemovich
 * Resolved #283, improved object resolution of cypher_query so that it 
   resolves objects even if they are nested within lists. Overall 
   documentation edits - Athanasios Anastasiou
   
Version 3.2.9 2018-07-04
 * Add check for wiping db on test run - Athanasios 
 * Correct function name in doc string - Henry Jordan
 * Support filtering on properties that end with a hash (#348) - mprahl
 * Support neo4j-driver v1.6.0 (#347) - mprahl
 * Explicit write transaction mode (#337) - Robert Grant
 * Adds a check for Traversal's definition argument (#333) - Frank Sachsenheim

Version 3.2.8 2018-04-29
 * Fix syntax error in import

Version 3.2.7 2018-04-23
 * Back compat for exceptions module move - Robin Edwards
 * Lower the required pytz version (#328) - mprahl
 * Add the "disconnect_all" and "replace" methods on relationship properties (#327) - mprahl
 * Add the "first" and "first_or_none" methods on the NodeSet class (#325) - mprahl

Version 3.2.6 2018-04-04
 * Code clean ups and documentation improvments - Frank Sachsenheim
 * neomodel.properties.NormalClass is renamed to NormalizedClass
 * Respect when db_property is set in install_labels and get_or_create - mprahl
 * Use a clearer relationship name in the Relationships documentation - mprahl
 * Improve relationship documentation - mprahl
 * Ensure_connection only called when accessing the database, not when
 constructing the transaction decorator - Robert Grant
 * Test against newer neo4j releases - Frank Sachsenheim
 * Fixes and simplifies pickling of DoesNotExist subclasses - Frank Sachsenheim
 * Add NeomodelException and move exceptions to a module - Frank Sachsenheim
 * Many documentation improvements - Frank Sachsenheim
 * Add newline to is abstract warning #305 - Omer Yampel
 * Refactors tests to use pytest as runner.- Frank Sachsenheim
 * Remove support for Python 3.3 - Frank Sachsenheim

neomodel/core.py
import re
import sys
import warnings

from neomodel import config
from neomodel.exceptions import DoesNotExist, ClassAlreadyDefined
from neomodel.hooks import hooks
from neomodel.properties import Property, PropertyManager
from neomodel.util import Database, classproperty, _UnsavedNode, _get_node_properties
from neo4j.exceptions import ClientError

db = Database()


def drop_constraints(quiet=True, stdout=None):
    """
    Discover and drop all constraints.

    :type: bool
    :return: None
    """
    if not stdout or stdout is None:
        stdout = sys.stdout

    results, meta = db.cypher_query("CALL db.constraints()")
    pattern = re.compile(":(.*) \).*\.(\w*)")
    for constraint in results:
        # Versions prior to 4.0 have a very different return format
        if constraint[0].startswith("CONSTRAINT "):
            db.cypher_query("DROP {!s}".format(constraint[0]))
            match = pattern.search(constraint[0])
        else:
            db.cypher_query("DROP CONSTRAINT {!s}".format(constraint[0]))
            match = pattern.search(constraint[1])
        stdout.write(
            """ - Droping unique constraint and index on label {0} with property {1}.\n""".format(
                match.group(1), match.group(2)
            )
        )
    stdout.write("\n")


def drop_indexes(quiet=True, stdout=None):
    """
    Discover and drop all indexes.

    :type: bool
    :return: None
    """
    if not stdout or stdout is None:
        stdout = sys.stdout

    results, meta = db.cypher_query("CALL db.indexes()")
    pattern = re.compile(":(.*)\((.*)\)")
    for index in results:
        # Versions prior to 4.0 have a very different return format
        if not isinstance(index[0], int) and index[0].startswith("INDEX "):
            db.cypher_query("DROP " + index[0])
            match = pattern.search(index[0])
            stdout.write(
                " - Dropping index on label {0} with property {1}.\n".format(
                    match.group(1), match.group(2)
                )
            )
        else:
            db.cypher_query("DROP INDEX " + index[1])
            stdout.write(
                " - Dropping index on label {0} with property {1}.\n".format(
                    index[7][0], index[8][0]
                )
            )
    stdout.write("\n")


def remove_all_labels(stdout=None):
    """
    Calls functions for dropping constraints and indexes.

    :param stdout: output stream
    :return: None
    """

    if not stdout:
        stdout = sys.stdout

    stdout.write("Droping constraints...\n")
    drop_constraints(quiet=False, stdout=stdout)

    stdout.write("Droping indexes...\n")
    drop_indexes(quiet=False, stdout=stdout)


def install_labels(cls, quiet=True, stdout=None):
    """
    Setup labels with indexes and constraints for a given class

    :param cls: StructuredNode class
    :type: class
    :param quiet: (default true) enable standard output
    :param stdout: stdout stream
    :type: bool
    :return: None
    """
    if not stdout or stdout is None:
        stdout = sys.stdout

    if not hasattr(cls, "__label__"):
        if not quiet:
            stdout.write(
                " ! Skipping class {0}.{1} is abstract\n".format(
                    cls.__module__, cls.__name__
                )
            )
        return

    for name, property in cls.defined_properties(aliases=False, rels=False).items():
        db_property = property.db_property or name
        if property.index:
            if not quiet:
                stdout.write(
                    " + Creating index {0} on label {1} for class {2}.{3}\n".format(
                        name, cls.__label__, cls.__module__, cls.__name__
                    )
                )
            try:
                db.cypher_query(
                    "CREATE INDEX on :{0}({1}); ".format(cls.__label__, db_property)
                )
            except ClientError as e:
                if str(e).lower().startswith("an equivalent index already exists"):
                    stdout.write("{0}\n".format(str(e)))
                else:
                    raise

        elif property.unique_index:
            if not quiet:
                stdout.write(
                    " + Creating unique constraint for {0} on label {1} for class {2}.{3}\n".format(
                        name, cls.__label__, cls.__module__, cls.__name__
                    )
                )
            try:
                db.cypher_query(
                    "CREATE CONSTRAINT "
                    "on (n:{0}) ASSERT n.{1} IS UNIQUE".format(
                        cls.__label__, db_property
                    )
                )
            except ClientError as e:
                if str(e).lower().startswith("an equivalent constraint already exists"):
                    stdout.write("{0}\n".format(str(e)))
                else:
                    raise


def install_all_labels(stdout=None):
    """
    Discover all subclasses of StructuredNode in your application and execute install_labels on each.
    Note: code most be loaded (imported) in order for a class to be discovered.

    :param stdout: output stream
    :return: None
    """

    if not stdout or stdout is None:
        stdout = sys.stdout

    def subsub(cls):  # recursively return all subclasses
        return cls.__subclasses__() + [
            g for s in cls.__subclasses__() for g in subsub(s)
        ]

    stdout.write("Setting up indexes and constraints...\n\n")

    i = 0
    for cls in subsub(StructuredNode):
        stdout.write("Found {0}.{1}\n".format(cls.__module__, cls.__name__))
        install_labels(cls, quiet=False, stdout=stdout)
        i += 1

    if i:
        stdout.write("\n")

    stdout.write("Finished {0} classes.\n".format(i))


class NodeMeta(type):
    def __new__(mcs, name, bases, namespace):
        namespace["DoesNotExist"] = type(name + "DoesNotExist", (DoesNotExist,), {})
        cls = super(NodeMeta, mcs).__new__(mcs, name, bases, namespace)
        # needed by Python < 3.5 for unpickling DoesNotExist objects:
        cls.DoesNotExist._model_class = cls

        if hasattr(cls, "__abstract_node__"):
            delattr(cls, "__abstract_node__")
        else:
            if "deleted" in namespace:
                raise ValueError(
                    "Class property called 'deleted' conflicts "
                    "with neomodel internals."
                )
            for key, value in (
                (x, y) for x, y in namespace.items() if isinstance(y, Property)
            ):
                value.name, value.owner = key, cls
                if hasattr(value, "setup") and callable(value.setup):
                    value.setup()

            # cache various groups of properies
            cls.__required_properties__ = tuple(
                name
                for name, property in cls.defined_properties(
                    aliases=False, rels=False
                ).items()
                if property.required or property.unique_index
            )
            cls.__all_properties__ = tuple(
                cls.defined_properties(aliases=False, rels=False).items()
            )
            cls.__all_aliases__ = tuple(
                cls.defined_properties(properties=False, rels=False).items()
            )
            cls.__all_relationships__ = tuple(
                cls.defined_properties(aliases=False, properties=False).items()
            )

            cls.__label__ = namespace.get("__label__", name)

            if config.AUTO_INSTALL_LABELS:
                install_labels(cls, quiet=False)

            label_set = frozenset(cls.inherited_labels())
            if label_set not in db._NODE_CLASS_REGISTRY:
                db._NODE_CLASS_REGISTRY[label_set] = cls
            else:
                raise ClassAlreadyDefined(cls, db._NODE_CLASS_REGISTRY)

        return cls


NodeBase = NodeMeta("NodeBase", (PropertyManager,), {"__abstract_node__": True})


class StructuredNode(NodeBase):
    """
    Base class for all node definitions to inherit from.

    If you want to create your own abstract classes set:
        __abstract_node__ = True
    """

    # static properties

    __abstract_node__ = True

    # magic methods

    def __init__(self, *args, **kwargs):
        if "deleted" in kwargs:
            raise ValueError("deleted property is reserved for neomodel")

        for key, val in self.__all_relationships__:
            self.__dict__[key] = val.build_manager(self, key)

        super(StructuredNode, self).__init__(*args, **kwargs)

    def __eq__(self, other):
        if not isinstance(other, (StructuredNode,)):
            return False
        if hasattr(self, "id") and hasattr(other, "id"):
            return self.id == other.id
        return False

    def __ne__(self, other):
        return not self.__eq__(other)

    def __repr__(self):
        return "<{0}: {1}>".format(self.__class__.__name__, self)

    def __str__(self):
        return repr(self.__properties__)

    # dynamic properties

    @classproperty
    def nodes(cls):
        """
        Returns a NodeSet object representing all nodes of the classes label
        :return: NodeSet
        :rtype: NodeSet
        """
        from .match import NodeSet

        return NodeSet(cls)

    @property
    def _id(self, val):
        raise NotImplementedError(
            "the _id property was deprecated and has now been removed. "
            "Please use .id instead."
        )

    # methods

    @classmethod
    def _build_merge_query(
        cls, merge_params, update_existing=False, lazy=False, relationship=None
    ):
        """
        Get a tuple of a CYPHER query and a params dict for the specified MERGE query.

        :param merge_params: The target node match parameters, each node must have a "create" key and optional "update".
        :type merge_params: list of dict
        :param update_existing: True to update properties of existing nodes, default False to keep existing values.
        :type update_existing: bool
        :rtype: tuple
        """
        query_params = dict(merge_params=merge_params)
        n_merge = "n:{0} {{{1}}}".format(
            ":".join(cls.inherited_labels()),
            ", ".join(
                "{0}: params.create.{0}".format(getattr(cls, p).db_property or p)
                for p in cls.__required_properties__
            ),
        )
        if relationship is None:
            # create "simple" unwind query
            query = "UNWIND $merge_params as params\n MERGE ({0})\n ".format(n_merge)
        else:
            # validate relationship
            if not isinstance(relationship.source, StructuredNode):
                raise ValueError(
                    "relationship source [{0}] is not a StructuredNode".format(
                        repr(relationship.source)
                    )
                )
            relation_type = relationship.definition.get("relation_type")
            if not relation_type:
                raise ValueError(
                    "No relation_type is specified on provided relationship"
                )

            from .match import _rel_helper

            query_params["source_id"] = relationship.source.id
            query = "MATCH (source:{0}) WHERE ID(source) = $source_id\n ".format(
                relationship.source.__label__
            )
            query += "WITH source\n UNWIND $merge_params as params \n "
            query += "MERGE "
            query += _rel_helper(
                lhs="source",
                rhs=n_merge,
                ident=None,
                relation_type=relation_type,
                direction=relationship.definition["direction"],
            )

        query += "ON CREATE SET n = params.create\n "
        # if update_existing, write properties on match as well
        if update_existing is True:
            query += "ON MATCH SET n += params.update\n"

        # close query
        if lazy:
            query += "RETURN id(n)"
        else:
            query += "RETURN n"

        return query, query_params

    @classmethod
    def category(cls):
        raise NotImplementedError(
            "Category was deprecated and has now been removed, "
            "the functionality is now achieved using the {0}.nodes attribute".format(
                cls.__name__
            )
        )

    @classmethod
    def create(cls, *props, **kwargs):
        """
        Call to CREATE with parameters map. A new instance will be created and saved.

        :param props: dict of properties to create the nodes.
        :type props: tuple
        :param lazy: False by default, specify True to get nodes with id only without the parameters.
        :type: bool
        :rtype: list
        """

        if "streaming" in kwargs:
            warnings.warn(
                "streaming is not supported by bolt, please remove the kwarg",
                category=DeprecationWarning,
                stacklevel=1,
            )

        lazy = kwargs.get("lazy", False)
        # create mapped query
        query = "CREATE (n:{0} $create_params)".format(":".join(cls.inherited_labels()))

        # close query
        if lazy:
            query += " RETURN id(n)"
        else:
            query += " RETURN n"

        results = []
        for item in [
            cls.deflate(p, obj=_UnsavedNode(), skip_empty=True) for p in props
        ]:
            node, _ = db.cypher_query(query, {"create_params": item})
            results.extend(node[0])

        nodes = [cls.inflate(node) for node in results]

        if not lazy and hasattr(cls, "post_create"):
            for node in nodes:
                node.post_create()

        return nodes

    @classmethod
    def create_or_update(cls, *props, **kwargs):
        """
        Call to MERGE with parameters map. A new instance will be created and saved if does not already exists,
        this is an atomic operation. If an instance already exists all optional properties specified will be updated.

        Note that the post_create hook isn't called after create_or_update

        :param props: List of dict arguments to get or create the entities with.
        :type props: tuple
        :param relationship: Optional, relationship to get/create on when new entity is created.
        :param lazy: False by default, specify True to get nodes with id only without the parameters.
        :rtype: list
        """
        lazy = kwargs.get("lazy", False)
        relationship = kwargs.get("relationship")

        # build merge query, make sure to update only explicitly specified properties
        create_or_update_params = []
        for specified, deflated in [
            (p, cls.deflate(p, skip_empty=True)) for p in props
        ]:
            create_or_update_params.append(
                {
                    "create": deflated,
                    "update": dict(
                        (k, v) for k, v in deflated.items() if k in specified
                    ),
                }
            )
        query, params = cls._build_merge_query(
            create_or_update_params,
            update_existing=True,
            relationship=relationship,
            lazy=lazy,
        )

        if "streaming" in kwargs:
            warnings.warn(
                "streaming is not supported by bolt, please remove the kwarg",
                category=DeprecationWarning,
                stacklevel=1,
            )

        # fetch and build instance for each result
        results = db.cypher_query(query, params)
        return [cls.inflate(r[0]) for r in results[0]]

    def cypher(self, query, params=None):
        """
        Execute a cypher query with the param 'self' pre-populated with the nodes neo4j id.

        :param query: cypher query string
        :type: string
        :param params: query parameters
        :type: dict
        :return: list containing query results
        :rtype: list
        """
        self._pre_action_check("cypher")
        params = params or {}
        params.update({"self": self.id})
        return db.cypher_query(query, params)

    @hooks
    def delete(self):
        """
        Delete a node and it's relationships

        :return: True
        """
        self._pre_action_check("delete")
        self.cypher(
            "MATCH (self) WHERE id(self)=$self "
            "OPTIONAL MATCH (self)-[r]-()"
            " DELETE r, self"
        )
        delattr(self, "id")
        self.deleted = True
        return True

    @classmethod
    def get_or_create(cls, *props, **kwargs):
        """
        Call to MERGE with parameters map. A new instance will be created and saved if does not already exist,
        this is an atomic operation.
        Parameters must contain all required properties, any non required properties with defaults will be generated.

        Note that the post_create hook isn't called after get_or_create

        :param props: Arguments to get_or_create as tuple of dict with property names and values to get or create
                      the entities with.
        :type props: tuple
        :param relationship: Optional, relationship to get/create on when new entity is created.
        :param lazy: False by default, specify True to get nodes with id only without the parameters.
        :rtype: list
        """
        lazy = kwargs.get("lazy", False)
        relationship = kwargs.get("relationship")

        # build merge query
        get_or_create_params = [
            {"create": cls.deflate(p, skip_empty=True)} for p in props
        ]
        query, params = cls._build_merge_query(
            get_or_create_params, relationship=relationship, lazy=lazy
        )

        if "streaming" in kwargs:
            warnings.warn(
                "streaming is not supported by bolt, please remove the kwarg",
                category=DeprecationWarning,
                stacklevel=1,
            )

        # fetch and build instance for each result
        results = db.cypher_query(query, params)
        return [cls.inflate(r[0]) for r in results[0]]

    @classmethod
    def inflate(cls, node):
        """
        Inflate a raw neo4j_driver node to a neomodel node
        :param node:
        :return: node object
        """
        # support lazy loading
        if isinstance(node, int):
            snode = cls()
            snode.id = node
        else:
            node_properties = _get_node_properties(node)
            props = {}
            for key, prop in cls.__all_properties__:
                # map property name from database to object property
                db_property = prop.db_property or key

                if db_property in node_properties:
                    props[key] = prop.inflate(node_properties[db_property], node)
                elif prop.has_default:
                    props[key] = prop.default_value()
                else:
                    props[key] = None

            snode = cls(**props)
            snode.id = node.id

        return snode

    @classmethod
    def inherited_labels(cls):
        """
        Return list of labels from nodes class hierarchy.

        :return: list
        """
        return [
            scls.__label__
            for scls in cls.mro()
            if hasattr(scls, "__label__") and not hasattr(scls, "__abstract_node__")
        ]

    def labels(self):
        """
        Returns list of labels tied to the node from neo4j.

        :return: list of labels
        :rtype: list
        """
        self._pre_action_check("labels")
        return self.cypher("MATCH (n) WHERE id(n)=$self " "RETURN labels(n)")[0][0][0]

    def _pre_action_check(self, action):
        if hasattr(self, "deleted") and self.deleted:
            raise ValueError(
                "{0}.{1}() attempted on deleted node".format(
                    self.__class__.__name__, action
                )
            )
        if not hasattr(self, "id"):
            raise ValueError(
                "{0}.{1}() attempted on unsaved node".format(
                    self.__class__.__name__, action
                )
            )

    def refresh(self):
        """
        Reload the node from neo4j
        """
        self._pre_action_check("refresh")
        if hasattr(self, "id"):
            request = self.cypher("MATCH (n) WHERE id(n)=$self" " RETURN n")[0]
            if not request or not request[0]:
                raise self.__class__.DoesNotExist("Can't refresh non existent node")
            node = self.inflate(request[0][0])
            for key, val in node.__properties__.items():
                setattr(self, key, val)
        else:
            raise ValueError("Can't refresh unsaved node")

    @hooks
    def save(self):
        """
        Save the node to neo4j or raise an exception

        :return: the node instance
        """

        # create or update instance node
        if hasattr(self, "id"):
            # update
            params = self.deflate(self.__properties__, self)
            query = "MATCH (n) WHERE id(n)=$self \n"
            query += "\n".join(
                ["SET n.{0} = ${1}".format(key, key) + "\n" for key in params.keys()]
            )
            for label in self.inherited_labels():
                query += "SET n:`{0}`\n".format(label)
            self.cypher(query, params)
        elif hasattr(self, "deleted") and self.deleted:
            raise ValueError(
                "{0}.save() attempted on deleted node".format(self.__class__.__name__)
            )
        else:  # create
            self.id = self.create(self.__properties__)[0].id
        return self
